core/src/main/scala/org/apache/spark/TaskEndReason.scala

```diff
@@ -127,6 +146,19 @@ case class ExceptionFailure(
   }
 }
 
+class ThrowableSerializationWrapper(var exception: Throwable) extends Serializable {
+  private def writeObject(out: ObjectOutputStream): Unit = {
+    out.writeObject(exception)
+  }
+  private def readObject(in: ObjectInputStream): Unit = {
+    try {
+      exception = in.readObject().asInstanceOf[Throwable]
+    } catch {
+      case _ : ClassNotFoundException | _ : IOException => // exception is null
```

Comment:
We could log the error at warning level so users know why they're not seeing the actual exception.